### PR TITLE
Implement byte array reusage in `NioTransport`

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #27563. This commit modifies the
`InboundChannelBuffer` to support releasable byte pages. These byte
pages are provided by the `PageCacheRecycler`. The `PageCacheRecycler`
must be passed to the `Transport` with this change.

### CODE REVIEW

**Comment by tbrooks8** [comment]
The suggestion to pass around the `PageCacheRecycler` came from this [comment]() by @jpountz.

Right now, I modified our generic tests cases (ESTestCase and ESIntegTestCase) to only check that the page cache recycler has released all pages `AfterClass`. We still check the `BigArrays` in between tests. This is because in integration tests a channel could hold byte arrays as long as the transport is running. I'm not sure of another way to solve this issue right now. But I am open to ideas.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java [hunk 1]
```diff
@@ -201,4 +213,24 @@ private int pageIndex(long index) {
...
+            this.byteBuffer = byteBuffer;
+            this.releasable = releasable;
+        }
+
+        @Override
+        public void close() {
+            releasable.close();
+        }
+
+        private ByteBuffer byteBuffer() {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java]: I think we can just access the member directly.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java [hunk 2]
```diff
@@ -30,39 +33,48 @@
...
+    private final ArrayDeque<Page> pages;
+    private final Supplier<Page> pageSupplier;
 
     private long capacity = 0;
     private long internalIndex = 0;
     // The offset is an int as it is the offset of where the bytes begin in the first buffer
     private int offset = 0;
 
     public InboundChannelBuffer() {
-        this(() -> ByteBuffer.wrap(new byte[PAGE_SIZE]));
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java]: can we remove this default ctor and pass the non-recycling closure in the test directly. I think it's just one place.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java [hunk 3]
```diff
@@ -30,39 +33,48 @@
...
         this.pages = new ArrayDeque<>();
         this.capacity = PAGE_SIZE * pages.size();
         ensureCapacity(PAGE_SIZE);
     }
 
+    @Override
+    public void close() {
+        Page page;
+        while ((page = pages.pollFirst()) != null) {
+            page.close();
```
[the remark this answers is not in the document]
Comment (s1monw) [reply test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java]: now thta we close should we prevent access after we are closed?

**Review by s1monw — COMMENTED** [review]
I left some minors. looks good

**Comment by tbrooks8** [comment]
@s1monw I've made changes based on your review.

**Comment by jpountz** [comment]
I like the approach!

File: test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java [hunk 4]
```diff
@@ -62,13 +60,30 @@ public InboundChannelBuffer(Supplier<Page> pageSupplier) {
...
-        Page page;
-        while ((page = pages.pollFirst()) != null) {
-            page.close();
+        if (isClosed.compareAndSet(false, true)) {
+            Page page;
+            RuntimeException closingException = null;
+            while ((page = pages.pollFirst()) != null) {
+                try {
+                    page.close();
+                } catch (RuntimeException e) {
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java]: you can use `ExceptionsHelper#rethrowAndSuppress` here

**Review by s1monw — APPROVED** [review]
LGTM - left one comment

**Pull request merged**